pandas/tests/groupby/test_groupby.py

```diff
@@ -2348,6 +2348,52 @@ def test_groupby_duplicate_index():
     tm.assert_series_equal(result, expected)
 
 
+def test_group_on_empty_multiindex(transformation_func, request):
+    # GH 47787
+    # With one row, those are transforms so the schema should be the same
+    if transformation_func == "tshift":
+        mark = pytest.mark.xfail(raises=NotImplementedError)
+        request.node.add_marker(mark)
+    df = DataFrame(
+        data=[[1, Timestamp("today"), 3, 4]],
+        columns=["col_1", "col_2", "col_3", "col_4"],
+    )
+    df = df.set_index(["col_1", "col_2"])
+    if transformation_func == "fillna":
+        args = ("ffill",)
+    elif transformation_func == "tshift":
+        args = (1, "D")
+    else:
+        args = ()
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(df.index, result.index)
+    has_same_columns = (
+        type(result) == DataFrame and (df.columns == result.columns).all()
+    )
+
+    col_3 = df["col_3"]
+    result = col_3.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(col_3.index, result.index)
+
+    # When empty, expect the same schema as well
+    df = DataFrame(data=[], columns=["col_1", "col_2", "col_3", "col_4"], dtype=int)
+    df = df.set_index(["col_1", "col_2"])
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    assert df.index.names == result.index.names
+    type_changes = {"pct_change", "rank"}
+    # pct change will return a float, so different from the original type
+    if has_same_columns and transformation_func not in type_changes:
```

Comment:
For type_changes, can you do

```
expected = df.astype(float)
tm.assert_frame_equal(result, expected)
```

Also, I think a good way of structuring something like this is

```
if transformation_func in type_chanages:
    expected = df.astype(float)
else:
    expected = df
tm.assert_frame_equal(result, expected)
```

Which ops have `has_same_columns` False and can we test those?

Reply:
I was aiming to be as future proof as possible by generating 1 dataset with one row and ensuring consistency when there is no row. It's better to test a property (here: the schema should be the same whether there is one row or None) than to enumerate every transformation and hardcode the result, because the next transformation to pop up will not be taken into account by default. The ops that are problematic are ell the cum... ones + the ones explicitly mentioned in the test.

I don't think casting as float is the right thing. This is an artifact of not having a NaN int for this particular column, not a property of the transformation.

There is a lot of scope creep here: the behaviour I am trying to solve really is what happens for empty dataframes compared to non empty data frames. I don't want to write tests for each individual transformation, as this just makes the lib longer term harder to maintain.